feat(hicache): support numa detect to reduce long tail latency

xiaguan opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

Hi from [novita.ai](https://novita.ai/) team

## Motivation

For machines with multi-GPU, multi-NIC, and multi-NUMA setups, topology is super important.

It doesn't just slightly hurt performance—it can significantly impact long-tail latency (e.g., p99 and max).

In this PR, I added --hicache-numa-detect (disabled by default) to bind the TP workers to the correct NUMA node (both CPU and memory).

I also updated the benchmark script to report p99 and max latency, so we can clearly see long-tail issues.

Most of the performance gain comes from RDMA transfers—now we avoid cross-NUMA RDMA traffic entirely.

Here are the benchmark results: H800 × 8, 200Gbps RDMA × 8, TP=8, Qwen-235B with page size 64.

not set `--hicache-numa-detect`

```
All requests completed
Performance metrics summary:
  Total requests: 800 at 16.0 requests per second
  Average Prompt Length: 11231.94 tokens
  Average Output Length: 1.00 tokens
  Average TTFT: 0.86
  P90 TTFT: 0.77
  P99 TTFT: 14.59
  Median TTFT: 0.62
  Max TTFT: 18.98
  Average latency: 0.86
  P90 latency: 0.77
  P99 latency: 14.59
  Median latency: 0.62
  Max latency: 18.98
  Input token throughput: 49544.09 tokens per second
  Output token throughput: 4.41 tokens per second
  Request Throughput: 4.41 requests per second
  Cache Hit Rate: 0.794541
```
The hit ratio dropped slightly due to Mooncake's lease mechanism. Long-tail transfer times can exceed the lease duration, causing read failures.

Same machine, same setup script with `--hicache-numa-detect`

```
All requests completed
Performance metrics summary:
  Total requests: 800 at 16.0 requests per second
  Average Prompt Length: 11231.92 tokens
  Average Output Length: 1.00 tokens
  Average TTFT: 0.60
  P90 TTFT: 0.67
  P99 TTFT: 0.74
  Median TTFT: 0.61
  Max TTFT: 0.83
  Average latency: 0.60
  P90 latency: 0.67
  P99 latency: 0.74
  Median latency: 0.61
  Max latency: 0.83
  Input token throughput: 70869.97 tokens per second
  Output token throughput: 6.31 tokens per second
  Request Throughput: 6.31 requests per second
  Cache Hit Rate: 0.815939
```

<!-- Describe the purpose and goals of this pull request. -->

## Modifications

<!-- Detail the changes made in this pull request. -->

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

huangtingwei9988 commented:
> cc @huangtingwei9988

@xiaguan Thank you so much！！

xiaguan committed:
feat(hicache): support numa detect to reduce long tail latency

xiaguan committed:
Enhance ServerArgs to enable NUMA detection by default and clean up MooncakeStore imports and logic for batch operations.

xiaguan committed:
update import

xiaguan committed:
default true arg

xiaguan committed:
Refactor HiRadixCache and common utility: remove unnecessary whitespace and improve code readability.

xiaguan committed:
better handle numa detect for non cuda

xiaguan committed:
Merge branch 'main' into hicache_numa

xiaguan committed:
Refactor hierarchical cache NUMA detection configuration

- Renamed `hicache_numa_detect` to `disable_hicache_numa_detect` for clarity.
- Updated command-line argument to reflect the new naming and functionality.
- Adjusted logic in `HiRadixCache` to use the new configuration.
- Enhanced documentation in `get_physical_device_id` to clarify behavior with `CUDA_VISIBLE_DEVICES`.

xiaguan committed:
Update NUMA allocation method in `numa_bind_to_node` function

- Replaced `libnuma.numa_set_localalloc()` with `libnuma.numa_set_preferred(ctypes.c_int(node))` to improve memory allocation behavior on specified NUMA nodes.

xiaguan committed:
fix lint

xiezhq-hermann committed:
Merge branch 'main' into hicache_numa

hzh0425 reviewed (approved):
LGTM

xiaguan commented on python/sglang/srt/utils/common.py:
```diff
@@ -3773,3 +3780,62 @@ def get_or_create_event_loop():
+        )
+
+    try:
+        numa_id_str = output_line[len(prefix) :].strip()
```
I believe nvidia-smi -C only returns a single integer?
```
[-C | --get-numa-id-of-nearby-cpu]: Displays NUMA ID of the nearest CPU for a GPU
                 Used in conjunction with -i which must be a single device ID.
```

hzh0425 replied:
> The issue couldn't be reproduced on my H2O, perhaps due to different CPU platforms.
> 
> I'll apply a simple fix by just taking the first one.

A better approach might be to bind the first four GPUs to NUMA0 and the last four to NUMA1.

xiaguan replied:
Fixed

xiaguan committed:
Refactor NUMA ID extraction in `get_current_device_numa_node` function

- Updated the method to parse the NUMA ID by replacing commas with hyphens and extracting the first node, improving robustness in handling output format.

xiaguan committed:
Add functions to retrieve NUMA node count and system GPU count

- Introduced `get_numa_node_count` to determine the number of NUMA nodes available on the system.
- Added `get_system_gpu_count` to count the total number of physical GPUs, independent of CUDA visibility.
- Refactored `get_current_device_numa_node` to distribute GPUs across NUMA nodes based on GPU ID, improving resource allocation logic.

xiaguan committed:
fmt

xiaguan committed:
skip when numa is unavliable

xiaguan committed:
fix numa import issue

xiaguan committed:
revert fast path

xiaguan committed:
better handle

xiaguan committed:
Merge branch 'main' into hicache_numa

xiaguan committed:
try fix lint

xiezhq-hermann commented:
/rerun-failed-ci

xiezhq-hermann committed:
Merge branch 'main' into hicache_numa
